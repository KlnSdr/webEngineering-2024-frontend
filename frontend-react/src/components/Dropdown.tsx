import React, { useState } from "react";

interface DropdownProps {
  options: string[];
  initialValue: string;
  onChange: (value: string) => void;
}

const Dropdown: React.FC<DropdownProps> = ({
  options,
  initialValue,
  onChange,
}) => {
  const [value, setValue] = useState(initialValue);
  return (
    <select
      onChange={(e) => {
        setValue(e.target.value);
        onChange(e.target.value);
      }}
    >
      {options.map((option: string) => {
        if (option === value) {
          return (
            <option value={option} selected>
              {option}
            </option>
          );
        } else {
            return <option value={option}>{option}</option>;
        }
      })}
    </select>
  );
};

export default Dropdown;
