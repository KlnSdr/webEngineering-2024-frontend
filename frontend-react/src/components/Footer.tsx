import React from "react";
import Stack from "react-bootstrap/Stack";

interface FooterProps {
  owner: string;
  timestamp: string;
}

const Footer: React.FC<FooterProps> = ({ owner, timestamp }) => {
    return(
        <footer className="text-center bg-body-tertiary pt-3">
            <Stack direction={"horizontal"} gap={5}>
                <text> Erstellt von: {owner} </text>
                <text> Datum: {timestamp} </text>
            </Stack>
        </footer>
    );
}

export default Footer;