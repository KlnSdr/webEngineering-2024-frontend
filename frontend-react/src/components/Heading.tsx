import React from "react";

interface HeadingProps {
  headingText: string;
}

const Heading: React.FC<HeadingProps> = ({ headingText }) => {
  return <h1 className="headline-with-lines">{headingText}</h1>;
};

export default Heading;
