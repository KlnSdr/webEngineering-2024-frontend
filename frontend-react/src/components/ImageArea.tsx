import React from "react";


interface ImageSrc{
    origin: string;
}

const ImageArea: React.FC<ImageSrc> = ({origin}) => {
  return (
    <div className="ImageArea">
      <img src={origin} alt="Image"/>
    </div>
  );
}

export default ImageArea;