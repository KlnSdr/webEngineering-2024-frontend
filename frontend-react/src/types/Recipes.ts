import { NeededProduct } from "./Products";

interface CreateRecipe {
  title: string;
  image: File | null;
  description: string;
  products: NeededProduct[];
}

export type { CreateRecipe };
