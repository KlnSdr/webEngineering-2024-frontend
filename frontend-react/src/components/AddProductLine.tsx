import React, { useState } from "react";
import Dropdown from "./Dropdown";
import Button from "react-bootstrap/Button";
import Stack from "react-bootstrap/Stack";

interface AddProductLineProps {
  products: string[];
  initialValue: string;
  onChange: (product: string, amount: number) => void;
  onRemove: () => void;
  getUnitOf: (product: string) => string;
}

const AddProductLine: React.FC<AddProductLineProps> = ({
  products,
  initialValue,
  onChange,
  onRemove,
  getUnitOf,
}) => {
  const [unit, setUnit] = useState(getUnitOf(initialValue));
  const [amount, setAmount] = useState(0);
  const [product, setProduct] = useState(initialValue);

  return (
    <Stack direction="horizontal">
      <Dropdown
        options={products}
        initialValue={product}
        onChange={(val: string) => {
          setProduct(val);
          setUnit(getUnitOf(val));
          onChange(val, amount);
        }}
      />
      <input
        type="number"
        value={amount}
        onChange={(e) => {
          setAmount(parseInt(e.target.value));
          onChange(product, parseInt(e.target.value));
        }}
        className="form-control"
      />
      <label>{unit || getUnitOf(product)}</label>
      <Button
        onClick={() => {
          onRemove();
        }}
        className="bi bi-x"
      ></Button>
    </Stack>
  );
};

export default AddProductLine;
