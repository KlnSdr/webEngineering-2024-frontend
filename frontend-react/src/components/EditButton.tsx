import React from "react";
import {CreateRecipe} from "../types/Recipes";

interface EditButtonProps {
    CreateRecipe: CreateRecipe;
}

const EditButton: React.FC<EditButtonProps> = ({CreateRecipe}) => {
    return (
        <div className="EditButton">
            <button className="bi bi-pencil" onClick={() => console.log(CreateRecipe)}>Edit</button>
        </div>
    );
};

export default EditButton;




