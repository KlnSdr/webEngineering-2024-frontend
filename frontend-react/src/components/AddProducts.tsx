import React, { useState } from "react";
import AddProductLine from "./AddProductLine";
import Button from "react-bootstrap/Button";
import { NeededProduct, Product } from "../types/Products";

interface AddProductsProps {
  onChange: (products: NeededProduct[]) => void;
  availableProducts: Product[];
}

const AddProducts: React.FC<AddProductsProps> = ({
  onChange,
  availableProducts,
}) => {
  const [neededProducts, setNeededProducts] = useState<NeededProduct[]>([]);
  const emptyNeededProduct: NeededProduct = {
    id: -1,
    productName: "",
    amount: 0,
  };

  const updateNeededProduct = (
    index: number,
    productName: string,
    amount: number
  ) => {
    const neededProductsCopy: NeededProduct[] = [...neededProducts];
    neededProductsCopy[index] = {
      ...neededProductsCopy[index],
      productName: productName,
      amount: amount,
    };
    setNeededProducts(neededProductsCopy);
    onChange(neededProductsCopy);
  };

  const removeNeededProductAt = (index: number) => {
    const filtered = neededProducts.filter((_, i: number) => i !== index);
    setNeededProducts(filtered);
    onChange(filtered);
  };

  return (
    <div>
      {neededProducts.map((product: NeededProduct, index: number) => {
        return (
          <AddProductLine
            key={product.id}
            products={availableProducts.map((prod: Product) => prod.name)}
            initialValue={product.productName}
            onChange={(prod: string, amount: number) => {
              updateNeededProduct(index, prod, amount);
            }}
            onRemove={() => {
              removeNeededProductAt(index);
            }}
            getUnitOf={(product: string) => {
              return (
                availableProducts.find((prod: Product) => prod.name === product)
                  ?.unit || ""
              );
            }}
          />
        );
      })}

      <Button
        onClick={() => {
          const neededProductsNew: NeededProduct[] = [
            ...neededProducts,
            { ...emptyNeededProduct, id: Date.now() },
          ];
          setNeededProducts(neededProductsNew);
          onChange(neededProductsNew);
        }}
        className="bi bi-plus"
      ></Button>
    </div>
  );
};

export default AddProducts;
