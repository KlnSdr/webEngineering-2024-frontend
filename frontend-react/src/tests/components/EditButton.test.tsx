import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import EditButton from "../../components/EditButton";
import {CreateButton} from "../../components/EditButton";
import {Recipe} from "../../types/Recipes";
import {useNavigate} from "react-router-dom";

describe("EditButton Component", () => {
    const recipe: Recipe = {
        id: 1,
        title: "Test Recipe",
        description: "Test Description",
        imgUri: "test.jpg",
        isPrivate: false,
        creationDate: new Date(),
        ownerUri: "test",
        likedByUserUris: [],
        products:[]

    };

    test("logs the correct recipe on click", () => {
        console.log = jest.fn();

        const {asFragment} = render(<EditButton Recipe={recipe} />);

        const buttonElement = screen.getByRole("button");
        fireEvent.click(buttonElement);

        expect(console.log).toHaveBeenCalledWith(recipe);
        expect(asFragment()).toMatchSnapshot();
    });
});

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: jest.fn(),
}));

describe("CreateButton Component", () => {
    test("navigates to the correct link on click", () => {
        const navigate = jest.fn();
        (useNavigate as jest.Mock).mockReturnValue(navigate);

        const { asFragment } = render(<CreateButton Link={"./new"} />);

        const buttonElement = screen.getByRole("button");
        fireEvent.click(buttonElement);

        expect(navigate).toHaveBeenCalledWith("./new");
        expect(asFragment()).toMatchSnapshot();
    });

    test("renders the correct text", () => {
        render(<CreateButton Link={"./new"} />);
        expect(screen.getByText("Erstelle neue Umfrage")).toBeInTheDocument();
    });
});
