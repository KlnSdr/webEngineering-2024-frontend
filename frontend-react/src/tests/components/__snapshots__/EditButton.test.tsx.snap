// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateButton Component navigates to the correct link on click 1`] = `
<DocumentFragment>
  <div
    class="CreateButton"
  >
    <button
      class="bi bi-plus btn btn-primary"
      type="button"
    >
      Erstelle neue Umfrage
    </button>
  </div>
</DocumentFragment>
`;

exports[`EditButton Component logs the correct recipe on click 1`] = `
<DocumentFragment>
  <div
    class="EditButton"
  >
    <button
      class="bi bi-pencil btn btn-primary"
      type="button"
    >
      Edit
    </button>
  </div>
</DocumentFragment>
`;
